### PR TITLE
feat(utxoSelect): use `input.from` as change address if passed

### PR DESCRIPTION
**Pull request by gomesalexandre**
#### Description

If `input.from` is passed, the intent is to escape hatch the xpub privacy guarantees for destination addresses which business logic expect a single input address e.g THORSwap.

This PR makes sure we not only filter out UTXOs when `from` is passed, but also use it as a change address. This is effectively mimicking the behavior of wallets using a single *address* for UTXOs e.g THORSwap vs `wallets` using xpubs e.g us.

If the change address is automatically selected to be the next receive address, then this means either the `from` flow would work only once, or a Tx would have to be made to the previously used `from` address to populate an UTXO with the right value.

https://live.blockcypher.com/btc/tx/262848622bf782971dffa316dc755e9a2fac24555c4d847ce9356a889c0488f8/

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1168** <a href="https://app.graphite.dev/github/pr/shapeshift/lib/1168" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/lib/1168?utm_source=stack-comment).

File: packages/chain-adapters/src/utxo/utxoSelect/index.ts [hunk 1]
```diff
@@ -44,5 +44,14 @@ export const utxoSelect = (input: UTXOSelectInput) => {
     return coinSelect<unchained.bitcoin.Utxo>(utxos, outputs, Number(input.satoshiPerByte))
   })()
 
+  if (input.from && result?.outputs?.[1]?.value) {
+    // If input contains a `from` param, inputs will be filtered to only keep UTXOs from that address
+    // The change address will be set to this from address, so that it can be reused
+    // Reusing addresses is an xpub antipattern and totally voids UTXO privacy guarantees, thus input.from should only be used
+    // when dealing with destinations that expect address reuse e.g THORSwap
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/utxo/utxoSelect/index.ts:51]:
```suggestion
    // when dealing with destinations that expect address reuse e.g THORChain savers
```

File: packages/chain-adapters/src/utxo/utxoSelect/index.ts [hunk 2]
```diff
@@ -44,5 +44,14 @@ export const utxoSelect = (input: UTXOSelectInput) => {
     return coinSelect<unchained.bitcoin.Utxo>(utxos, outputs, Number(input.satoshiPerByte))
   })()
 
+  if (input.from && result?.outputs?.[1]?.value) {
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/utxo/utxoSelect/index.ts:47]:
```suggestion
  /**
   * note - the 0-indexed output is the `to` address, and if `input.from` is included as an argument, the
   * 1-indexed output will be the change address, which we overwrite below
   */
  if (input.from && result?.outputs?.[1]?.value) {
```

Comment (0xdef1cafe) [reply packages/chain-adapters/src/utxo/utxoSelect/index.ts:47]: comments explaining implicit logic - this bit i had to scrutinize a little to understand what's going on. looks good tho!

**Review by 0xdef1cafe — APPROVED** [review]
approved with comments - feel free to address and force merge ser

**Pull request merged**